File: algo/15. 3Sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        # time complexity: O(n2)
        
        if len(nums) < 3:
            return []
        
        nums = sorted(nums)
        res = {}
        
        wc = {}
        for num in nums:
            wc[num] = wc.get(num, 0) + 1
        
        if wc.get(0) and wc.get(0) >= 3:
            res[(0, 0, 0)] = 1
        
        mid = 0
        for i in range(len(nums)):
            if nums[i]>=0:
                mid = i
                break
        
        for i in range(0, mid):
            left = nums[i]
            for j in range(mid, len(nums)):
                right = nums[j]
                target = -left-right
                if wc.get(target):
                    if (wc[target] == 1) and ((target==left) or (target==right)):
                        continue
                    if tuple(sorted([left, right, target])) not in res.keys():
                        res[tuple(sorted([left, right, target]))] = 1
        return list(res.keys())
                        
                    
                

File: round3/15. 3Sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        var res = new List<IList<int>>();
        
        if (nums.Length < 3) return res;
        
        Array.Sort(nums);
        for (int i = 0; i < nums.Length - 2; i ++) {
            if (i != 0 && nums[i] == nums[i-1]) continue;
            int curr = nums[i];
            
            int left = i + 1;
            int right = nums.Length - 1;
            while (left < right) {
                if (nums[left] + nums[right] + curr == 0) {
                    res.Add(new List<int>(){curr, nums[left], nums[right]});
                    left ++;
                    right --;
                    while (left < nums.Length && nums[left] == nums[left - 1]) 
                        left++;
                    while (right >= 0 && nums[right] == nums[right + 1]) 
                        right--;
                } else if (nums[left] + nums[right] < -curr) {
                    left ++;
                } else {
                    right --;
                }
            } 
        }
        
        return res;        
    }
}